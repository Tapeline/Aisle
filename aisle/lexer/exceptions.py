"""Contains lexer exceptions."""


class LexerException(Exception):
    def __init__(self, message: str, source: str, line: int) -> None:
        self._msg = message
        self._src = source
        self._line = line

    @property
    def formatted_message(self):
        line_str = self._src.splitlines()[self._line - 1]
        line_str = f"{self._line} |  {line_str}"
        return (
            f"{self._msg}\n"
            f"At line {self._line}:\n"
            f"{line_str}"
        )

    def __str__(self):
        return self.formatted_message


class UnexpectedCharacterException(LexerException):
    def __init__(self, char: str, source: str, line: int) -> None:
        super().__init__(f"Unexpected character '{char}'", source, line)


class StringNotClosedException(LexerException):
    def __init__(self, source: str, line: int) -> None:
        super().__init__('String not closed with "', source, line)


class IncompleteUnicodeQuadException(LexerException):
    def __init__(self, source: str, line: int) -> None:
        super().__init__("\\u escape found, but code is incomplete", source, line)
