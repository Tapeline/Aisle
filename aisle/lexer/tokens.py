"""Contains token definitions."""

from dataclasses import dataclass
import enum
from typing import Final


class TokenType(enum.Enum):
    """All possible token types."""

    KEYWORD = enum.auto()
    TEXT = enum.auto()
    COLON = enum.auto()
    ASSIGN = enum.auto()
    ARROW_R = enum.auto()
    ARROW_L = enum.auto()
    ARROW_BI_DIR = enum.auto()
    ARROW_NO_DIR = enum.auto()
    LPAR = enum.auto()
    RPAR = enum.auto()
    LBRACKET = enum.auto()
    RBRACKET = enum.auto()
    INDENT = enum.auto()
    NEWLINE = enum.auto()


@dataclass
class Token:
    """Represents a single token."""

    type: TokenType
    lexeme: str
    line: int

    def is_keyword(self, pattern: str) -> bool:
        """Check if this token is a specific keyword."""
        return self.type == TokenType.KEYWORD and self.lexeme == pattern

    def __repr__(self):
        return f"{self.type.name}{self.lexeme}"


KEYWORDS: Final[set[str]] = {
    "scope",
    "context",
    "system",
    "external",
    "links",
    "over",
    "tech",
    "containers",
    "service",
    "deployment",
    "deploy",
}

RESERVED_STRINGS: Final[set[str]] = {
    *KEYWORDS,
    "-->",
    "<--",
    "<->",
    "---",
    *set("()[]:=")
}
