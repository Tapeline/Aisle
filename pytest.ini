[pytest]
addopts =
    --cov-fail-under=100
    --cov-report=term-missing:skip-covered
    --cov=aisle
    --cov=tests
