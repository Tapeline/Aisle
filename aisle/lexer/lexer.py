"""Contains lexer implementation."""

from collections.abc import Sequence

from aisle.lexer.exceptions import (
    UnexpectedCharacterException,
    StringNotClosedException,
    IncompleteUnicodeQuadException,
)
from aisle.lexer.tokens import (
    Token,
    TokenType,
    RESERVED_STRINGS,
    KEYWORDS,
)


class Lexer:
    def __init__(
            self,
            source: str,
            tab_size: int = 4,
    ) -> None:
        self._src = _norm_newlines(source)
        self._lines = self._src.splitlines()
        self._i = 0
        self._start = 0
        self._line = 1
        self._line_start = 0
        self._tokens = []
        self._tab_size = tab_size

    @property
    def _at_end(self):
        """Check if reached end of source."""
        return self._i >= len(self._src)

    def _next(self) -> str | None:
        if self._i >= len(self._src):
            return None
        self._i += 1
        return self._src[self._i - 1]

    @property
    def _current(self) -> str | None:
        return None if self._at_end else self._src[self._i]

    def _add_token(
            self,
            token_type: TokenType,
            lexeme: str | None = None
    ) -> None:
        token = Token(
            token_type,
            lexeme or self._src[self._start:self._i],
            self._line
        )
        self._start = self._i
        self._tokens.append(token)

    def _match(self, pattern: str) -> str | None:
        if self._at_end:
            return None
        src = self._src[self._i:]
        if src.startswith(pattern):
            self._i += len(pattern)
            return pattern
        return None

    def _error(self, exc_type, *args, **kwargs) -> None:
        raise exc_type(
            *args,
            **kwargs,
            source=self._src,
            line=self._line
        )

    def scan(self) -> Sequence[Token]:
        while not self._at_end:
            self._scan_token()
        return self._tokens

    def _scan_token(self) -> None:
        c = self._next()
        match c:
            case "=":
                self._add_token(TokenType.ASSIGN)
            case ":":
                self._add_token(TokenType.COLON)
            case "(":
                self._add_token(TokenType.LPAR)
            case ")":
                self._add_token(TokenType.RPAR)
            case "[":
                self._add_token(TokenType.LBRACKET)
            case "]":
                self._add_token(TokenType.RBRACKET)
            case "-":
                if self._match("->"):
                    self._add_token(TokenType.ARROW_R)
                elif self._match("--"):
                    self._add_token(TokenType.ARROW_NO_DIR)
                else:
                    self._error(UnexpectedCharacterException, self._current)
            case "<":
                if self._match("->"):
                    self._add_token(TokenType.ARROW_BI_DIR)
                elif self._match("--"):
                    self._add_token(TokenType.ARROW_L)
                else:
                    self._error(UnexpectedCharacterException, self._current)
            case "\n":
                self._add_token(TokenType.NEWLINE)
                self._line += 1
                self._line_start = self._i
            case " ":
                if self._match(" " * (self._tab_size - 1)):
                    self._add_token(TokenType.INDENT)
                else:
                    self._start = self._i
            case "\t":
                self._add_token(TokenType.INDENT)
            case '"':
                self._scan_string()
            case _:
                self._i -= 1
                self._scan_text()

    def _scan_string(self) -> None:
        text = ""
        while not self._at_end and self._current != '"':
            if self._current == "\\":
                text += self._scan_escape_seq()
            else:
                text += self._current
                self._i += 1
        if not (not self._at_end and self._current == '"'):
            self._error(StringNotClosedException)
        self._i += 1
        self._add_token(TokenType.TEXT, text)

    def _scan_escape_seq(self) -> str:
        self._next()
        control = self._next()
        match control:
            case "r":
                return "\r"
            case "f":
                return "\f"
            case "n":
                return "\n"
            case "t":
                return "\t"
            case '"':
                return '"'
            case "\\":
                return "\\"
            case "u":
                code = int(self._scan_unicode_quad(), 16)
                return chr(code)
        self._error(UnexpectedCharacterException, char=control)

    def _scan_unicode_quad(self) -> str:
        code = ""
        for _ in range(4):
            c = self._next()
            if c is None:
                self._error(IncompleteUnicodeQuadException)
            if c not in "0123456789abcdefABCDEF":
                self._error(IncompleteUnicodeQuadException)
            code += c
        return code

    def _scan_text(self) -> None:
        subindent_level = self._i - self._line_start
        text = ""
        while not self._at_end:
            if self._current == "\\":
                text += self._scan_escape_seq()
            elif self._current == "\n":
                self._line += 1
                if not self._next_line_on_same_level(subindent_level):
                    break
                text += self._current
                text += " "*subindent_level  # Retain all indentation
                self._i += subindent_level + 1
            else:
                text += self._current
                self._i += 1
            if _get_reserved_string_at_end(text):
                break
        reserved = _get_reserved_string_at_end(text)
        if reserved and len(reserved) != len(text):
            self._i -= len(reserved)
            text = text.removesuffix(reserved)

        # Return all extra indentation and newline symbols back
        stripped_text = text.strip()
        suffix_start = text.index(stripped_text) + len(stripped_text)
        suffix_length = len(text) - suffix_start
        self._i -= suffix_length
        # Update line count accordingly
        self._line -= text[suffix_start:].count("\n")
        # Remove unnecessary indentation
        text = _remove_indentation(text, subindent_level)

        if text in KEYWORDS:
            self._add_token(TokenType.KEYWORD, text.strip())
        else:
            self._add_token(TokenType.TEXT, text.strip())

    def _next_line_on_same_level(self, subindent_level: int) -> bool:
        return self._lines[self._line - 1].startswith(" " * subindent_level)


def _get_reserved_string_at_end(text: str) -> str | None:
    if text in RESERVED_STRINGS:
        return text
    for string in RESERVED_STRINGS:
        if _should_be_split(text, string):
            return string
    return None


def _should_be_split(string: str, reserved: str) -> bool:
    if not string.endswith(reserved):
        return False
    trunk = string.removesuffix(reserved)
    if not reserved.isalnum():
        return True
    return trunk[-1] in "\n\r\t "


def _norm_newlines(text: str) -> str:
    """Replace all CRLF with LF."""
    return text.replace("\r\n", "\n").replace("\r", "\n")


def _remove_indentation(text: str, amount: int) -> str:
    lines = text.splitlines()
    prefix = " "*amount
    for i, line in enumerate(lines):
        lines[i] = line.removeprefix(prefix)
    return "\n".join(lines)

"""
tech = fdffd

"""