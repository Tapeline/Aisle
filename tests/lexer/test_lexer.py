"""Contains lexer tests."""
from pathlib import Path
from typing import Final

import pytest

from aisle.lexer.lexer import Lexer
from aisle.lexer.tokens import Token, TokenType


@pytest.mark.parametrize(
    ("src", "expected"),
    [
        (
            ":=()[]",
            [
                "COLON:",
                "ASSIGN=",
                "LPAR(",
                "RPAR)",
                "LBRACKET[",
                "RBRACKET]",
            ],
        ),
        (
            "This_is_a_text",
            ["TEXTThis_is_a_text"],
        ),
        (
            "scope",
            ["KEYWORDscope"],
        ),
        (
            "---<----><->",
            [
                "ARROW_NO_DIR---",
                "ARROW_L<--",
                "ARROW_R-->",
                "ARROW_BI_DIR<->",
            ],
        ),
        (
            "\n\r\n\t    ",  # also check that CRLF is processed correctly
            [
                "NEWLINE\n",
                "NEWLINE\n",
                "INDENT\t",
                "INDENT    ",
            ],
        ),
    ],
)
def test_simple_tokens(src, expected):
    """Test that simple tokens are processed correctly."""
    lexer = Lexer(src)
    tokens = lexer.scan()
    assert list(map(str, tokens)) == expected


@pytest.mark.parametrize(
    ("src", "expected"),
    [
        (
            '"String"',
            ["TEXTString"],
        ),
        (
            '"String""Followed by a string"',
            [
                "TEXTString",
                "TEXTFollowed by a string",
            ],
        ),
        (
            '"Escape \\n \\r \\t \\f \\u0462 \\\\ \\""',
            ['TEXTEscape \n \r \t \f \u0462 \\ "'],
        ),
    ],
)
def test_strings(src, expected):
    """Test that strings are processed correctly."""
    lexer = Lexer(src)
    tokens = lexer.scan()
    assert list(map(str, tokens)) == expected


code_all_quoted: Final[str] = (
    'service Backend:\n'
    '    "Provides API\\nMaybe split into microservices later"\n'
    '    tech = "Litestar, sqlalchemy, \\nfaststream"\n'
    '    links:\n'
    '        --> "Anti-fraud Service" over "HTTP"'
)
code_unquoted: Final[str] = (
    "service Backend:\n"
    "    Provides API\\nMaybe split into microservices later\n"
    "    tech = Litestar, sqlalchemy, \\nfaststream\n"
    "    links:\n"
    "        --> Anti-fraud Service over HTTP"
)
code_unquoted_multiline: Final[str] = (
    "service Backend:\n"
    "    Provides API\n"
    "    Maybe split into microservices later\n"
    "    tech = Litestar, sqlalchemy, \n"
    "           faststream\n"
    "    links:\n"
    "        --> Anti-fraud Service over HTTP"
)
expected_tokens: Final[list[str]] = [
    "KEYWORDservice",
    "TEXTBackend",
    "COLON:",
    "NEWLINE\n",
    "INDENT    ",
    "TEXTProvides API\nMaybe split into microservices later",
    "NEWLINE\n",
    "INDENT    ",
    "KEYWORDtech",
    "ASSIGN=",
    "TEXTLitestar, sqlalchemy, \nfaststream",
    "NEWLINE\n",
    "INDENT    ",
    "KEYWORDlinks",
    "COLON:",
    "NEWLINE\n",
    "INDENT    ",
    "INDENT    ",
    "ARROW_R-->",
    "TEXTAnti-fraud Service",
    "KEYWORDover",
    "TEXTHTTP",
]


@pytest.mark.parametrize(
    ("src", "expected"),
    [
        (code_all_quoted, expected_tokens),
        (code_unquoted, expected_tokens),
        (code_unquoted_multiline, expected_tokens),
    ],
)
def test_complex_code(src, expected):
    """Test that a complex code is lexed correctly."""
    lexer = Lexer(src)
    tokens = lexer.scan()
    assert list(map(str, tokens)) == expected


def test_multiline_edge_case():
    """Cover """
    src = (
        "scope project Test\n"
        "scope context Test\n"
        "system TestSystem:\n"
        "    description\n"
        "    tech = Tech\n"
        "    links:\n"
        "        --> Test\n"
    )
    tokens = Lexer(src).scan()
    expected = [
        "KEYWORDscope",
        "KEYWORDproject",
        "TEXTTest",
        "NEWLINE\n",
        "KEYWORDscope",
        "KEYWORDcontext",
        "TEXTTest",
        "NEWLINE\n",
        "KEYWORDsystem",
        "TEXTTestSystem",
        "COLON:",
        "NEWLINE\n",
        "INDENT    ",
        "TEXTdescription",
        "NEWLINE\n",
        "INDENT    ",
        "KEYWORDtech",
        "ASSIGN=",
        "TEXTTech",
        "NEWLINE\n",
        "INDENT    ",
        "KEYWORDlinks",
        "COLON:",
        "NEWLINE\n",
        "INDENT    ",
        "INDENT    ",
        "ARROW_R-->",
        "TEXTTest",
        "NEWLINE\n"
    ]
    assert list(map(str, tokens)) == expected


def test_with_real_file(snapshot):
    """Test that real project declaration is lexed correctly."""
    with Path("./tests/fixtures/test_1.aisle").open("r") as src_file:
        src = src_file.read()
    lexer = Lexer(src)
    tokens = lexer.scan()
    tokens_str = "\n".join(map(str, tokens))
    assert tokens_str == snapshot


def test_token_human_readable_repr():
    token = Token(TokenType.KEYWORD, "system", 2)
    expected = "2\t| KEYWORD       system"
    assert token.human_readable_repr == expected
