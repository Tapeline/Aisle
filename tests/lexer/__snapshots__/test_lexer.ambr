# serializer version: 1
# name: test_with_real_file
  '''
  TEXTThis is a comment text
  NEWLINE
  
  NEWLINE
  
  KEYWORDscope
  KEYWORDproject
  TEXTBlogging app without description
  NEWLINE
  
  NEWLINE
  
  KEYWORDscope
  KEYWORDproject
  TEXTBlogging app
  COLON:
  NEWLINE
  
  INDENT    
  TEXTA simple blogging app architecture
  NEWLINE
  
  NEWLINE
  
  NEWLINE
  
  NEWLINE
  
  NEWLINE
  
  KEYWORDscope
  KEYWORDcontext
  TEXTBlogging app
  NEWLINE
  
  NEWLINE
  
  NEWLINE
  
  KEYWORDsystem
  TEXTBlogging System
  COLON:
  NEWLINE
  
  INDENT    
  TEXTStores and manages blogs
  NEWLINE
  
  INDENT    
  KEYWORDlinks
  COLON:
  NEWLINE
  
  INDENT    
  INDENT    
  ARROW_R-->
  TEXTMetrics System
  NEWLINE
  
  NEWLINE
  
  NEWLINE
  
  KEYWORDexternal
  KEYWORDsystem
  TEXTMetrics System
  COLON:
  NEWLINE
  
  INDENT    
  TEXTCollects site metrics
  We will use Yandex Metrics
  NEWLINE
  
  NEWLINE
  
  NEWLINE
  
  KEYWORDactor
  TEXTUser
  COLON:
  NEWLINE
  
  INDENT    
  TEXTUses blogging app. Reads and creates posts
  NEWLINE
  
  INDENT    
  KEYWORDlinks
  COLON:
  NEWLINE
  
  INDENT    
  INDENT    
  ARROW_R-->
  TEXTBlogging System
  COLON:
  NEWLINE
  
  INDENT    
  INDENT    
  INDENT    
  TEXTUses
  NEWLINE
  
  NEWLINE
  
  NEWLINE
  
  KEYWORDactor
  TEXTAdmin
  COLON:
  NEWLINE
  
  INDENT    
  TEXTBlog moderator
  NEWLINE
  
  INDENT    
  KEYWORDlinks
  COLON:
  NEWLINE
  
  INDENT    
  INDENT    
  ARROW_R-->
  TEXTBlogging System
  COLON:
  NEWLINE
  
  INDENT    
  INDENT    
  INDENT    
  TEXTUses
  NEWLINE
  
  INDENT    
  INDENT    
  ARROW_R-->
  TEXTMetrics System
  COLON:
  NEWLINE
  
  INDENT    
  INDENT    
  INDENT    
  TEXTReviews post metrics
  NEWLINE
  
  NEWLINE
  
  NEWLINE
  
  NEWLINE
  
  NEWLINE
  
  KEYWORDscope
  KEYWORDcontainers
  TEXTBlogging app
  NEWLINE
  
  NEWLINE
  
  NEWLINE
  
  KEYWORDservice
  TEXTFrontend Service
  COLON:
  NEWLINE
  
  INDENT    
  TEXTUI
  NEWLINE
  
  INDENT    
  KEYWORDsystem
  ASSIGN=
  TEXTBlogging System
  NEWLINE
  
  INDENT    
  KEYWORDtech
  ASSIGN=
  TEXTVite+Vue
  NEWLINE
  
  INDENT    
  KEYWORDlinks
  COLON:
  NEWLINE
  
  INDENT    
  INDENT    
  ARROW_R-->
  TEXTBackend Service
  KEYWORDover
  TEXTHTTP
  COLON:
  NEWLINE
  
  INDENT    
  INDENT    
  INDENT    
  TEXTSend API requests
  NEWLINE
  
  NEWLINE
  
  NEWLINE
  
  KEYWORDservice
  TEXTBackend
  COLON:
  NEWLINE
  
  INDENT    
  TEXTProvides API
  Maybe split into microservices later
  NEWLINE
  
  NEWLINE
  
  INDENT    
  KEYWORDsystem
  ASSIGN=
  TEXTBlogging System
  NEWLINE
  
  INDENT    
  KEYWORDtech
  ASSIGN=
  TEXTLitestar, sqlalchemy
  NEWLINE
  
  INDENT    
  KEYWORDlinks
  COLON:
  NEWLINE
  
  INDENT    
  INDENT    
  ARROW_R-->
  TEXTMetrics Service
  KEYWORDover
  TEXTHTTP
  COLON:
  NEWLINE
  
  INDENT    
  INDENT    
  INDENT    
  TEXTSend metrics info on each request
  NEWLINE
  
  INDENT    
  INDENT    
  ARROW_R-->
  TEXTDB Service
  COLON:
  NEWLINE
  
  INDENT    
  INDENT    
  INDENT    
  TEXTStores info
  NEWLINE
  
  NEWLINE
  
  NEWLINE
  
  KEYWORDservice
  TEXTDB Service
  COLON:
  NEWLINE
  
  INDENT    
  KEYWORDsystem
  ASSIGN=
  TEXTBlogging System
  NEWLINE
  
  INDENT    
  KEYWORDtech
  ASSIGN=
  TEXTPostgres
  NEWLINE
  
  NEWLINE
  
  NEWLINE
  
  KEYWORDexternal
  KEYWORDservice
  TEXTMetrics Service
  COLON:
  NEWLINE
  
  INDENT    
  KEYWORDsystem
  ASSIGN=
  TEXTMetrics System
  NEWLINE
  
  NEWLINE
  
  NEWLINE
  
  NEWLINE
  
  NEWLINE
  
  KEYWORDscope
  KEYWORDdeployment
  TEXTBlogging app
  NEWLINE
  
  NEWLINE
  
  NEWLINE
  
  KEYWORDdeployment
  TEXTVPS Deployment
  COLON:
  NEWLINE
  
  INDENT    
  TEXTDocker containers with Docker compose
  NEWLINE
  
  INDENT    
  TEXTDocker compose
  LBRACKET[
  TEXTdocker_compose
  RBRACKET]
  COLON:
  NEWLINE
  
  INDENT    
  INDENT    
  KEYWORDdeploy
  TEXTFrontend Service
  ASSIGN=
  TEXTDocker container
  NEWLINE
  
  INDENT    
  INDENT    
  KEYWORDdeploy
  TEXTBackend Service
  ASSIGN=
  TEXTDocker container
  NEWLINE
  
  INDENT    
  INDENT    
  KEYWORDdeploy
  TEXTDB Service
  ASSIGN=
  TEXTDocker container
  NEWLINE
  
  NEWLINE
  
  NEWLINE
  
  KEYWORDexternal
  KEYWORDdeployment
  TEXTProvided Metrics
  LBRACKET[
  TEXTYandex
  RBRACKET]
  COLON:
  NEWLINE
  
  INDENT    
  TEXTMetrics Service is already provided
  NEWLINE
  
  INDENT    
  KEYWORDdeploy
  TEXTMetrics Service
  ASSIGN=
  TEXTCloud provider
  NEWLINE
  
  NEWLINE
  
  NEWLINE
  
  NEWLINE
  
  NEWLINE
  
  KEYWORDscope
  KEYWORDlegend
  TEXTBlogging app
  NEWLINE
  
  NEWLINE
  
  NEWLINE
  
  LBRACKET[
  TEXTcontains
  TEXTDocker
  RBRACKET]
  COLON:
  NEWLINE
  
  INDENT    
  TEXTDocker container deployment
  NEWLINE
  
  INDENT    
  TEXTbg
  ASSIGN=
  TEXT#aaaaff
  NEWLINE
  
  NEWLINE
  
  NEWLINE
  
  LBRACKET[
  TEXTtag
  TEXTYandex
  RBRACKET]
  COLON:
  NEWLINE
  
  INDENT    
  TEXTbg
  ASSIGN=
  TEXT#777700
  NEWLINE
  
  NEWLINE
  
  NEWLINE
  
  LBRACKET[
  ASSIGN=
  TEXTVPS Deployment
  RBRACKET]
  COLON:
  NEWLINE
  
  INDENT    
  TEXTbg
  ASSIGN=
  TEXT#cccccc
  NEWLINE
  
  NEWLINE
  
  NEWLINE
  
  KEYWORDsystem
  COLON:
  NEWLINE
  
  INDENT    
  TEXTshape
  ASSIGN=
  TEXTrounded-rect
  NEWLINE
  
  '''
# ---
